Performance things

pauldraper opened the pull request:
* Avoid redundant .values calculation
* Cache .text and .values
* Deduplicate parameter values (#5)

I don't expect these to be very significant unless there are deep trees, or a large array is used as a parameter multiple times.

XeCycle commented:
Hi, thanks for your interest in this project.  Personally I am not working on any JS projects now; progress on this project can be slow.  As you can see it's already three years since my last commit.

As for this PR:
- "Fix build.js for new Node.js" can be a separate PR.  The callback should report errors, IMO.
- Can you add some tests for the values-reused case?

pauldraper commented:
Certainly. Honestly, I wanted to start with this to see if the project still had life before adding tests :)

If you are looking for help maintaining this, I'd be happy to do it.

pauldraper commented:
Alright

1. I do have to change strict-ssl to convince Travis to build.
2. Node.js didn't support Set/Map until 0.11.x, so I had to increase the version there, though testing on other versions can wait for another PR. (FYI, 6.x is the oldest maintained version.) 
3. I changed the way values are reused. Postgresql does type inference on the variables can only infer one type. So naive value reuse is not safe ``SQL`SELECT 0 WHERE 1 = ${null} AND '1' = ${null}` `` resulted in an error. (``SQL`SELECT 0 WHERE 1 = ${null}::int AND '1' = ${null}::text` `` is okay, but that's not the habitual usage.) So instead, value reuse happens only when a literal  is reused. So if you wanted reuse you'd have to ``var a = SQL`${null}`; SQL`SELECT 0 WHERE 1 = ${a} AND '1' = ${a}` ``.
4. I removed the caching which was avoiding duplicate work on `.text` and `.values`. On second thought, I wasn't sure it was worth it. I really just wanted to get rid of the bad big-O behavior on large trees, and that is still fixed. 
5. I added tests.

XeCycle commented:
Thank you.  If you would like to take over its maintaince, let me know your npm username.

pauldraper commented:
@XeCycle, Github and NPM usernames are the same: pauldraper.
